Add methods to manage first slide number(firstSlideNum)

hyun1024 opened the pull request:
<html>
<body>
<!--StartFragment--><html><head></head><body><h2 data-path-to-node="4" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Feature(XSLF): Add Methods to Manage Custom Starting Slide Number</h2><p data-path-to-node="5" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">This Pull Request introduces new user-friendly methods within <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">XMLSlideShow</code> to manage the custom starting slide number of a PowerPoint presentation, corresponding to the <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">firstSlideNum</code> attribute in <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">ppt/presentation.xml</code>.</p><h3 data-path-to-node="6" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Changes Summary</h3>
<li><code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">getFirstSlideNumber()</code>: Retrieves the current starting slide number (default is 1).
<li><code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">setFirstSlideNumber(int num)</code>: Sets the custom starting slide number.
<li><code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">unsetFirstSlideNumber()</code>: Removes the firstSlideNum attribute, reverting to the default (1).

<h3 data-path-to-node="8" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">🔍 Rationale for Adding High-Level API</h3><p data-path-to-node="9" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Although the underlying XMLBeans class (<code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">org.openxmlformats.schemas.presentationml.x2006.main.CTPresentation</code>) already contains low-level methods to manipulate <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">firstSlideNum</code>, we propose wrapping this functionality in <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">XMLSlideShow</code> for two critical reasons, aligning with POI's core design principles:</p><h4 data-path-to-node="10" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">1. Abstraction and API Usability</h4><p data-path-to-node="11" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">The first slide number is a standard, user-facing configuration option in Microsoft PowerPoint. Providing a high-level API prevents users from needing to manually retrieve and manipulate the internal <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">CTPresentation</code> object. This promotes cleaner, more readable code and shields developers from complex OOXML schema details.</p><h4 data-path-to-node="12" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">2. Enforcing Application Constraints (Data Validation)</h4><p data-path-to-node="13" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">The low-level <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">CTPresentation</code> methods adhere strictly to the general XML Schema definition (allowing any <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">int</code> value), but they <b style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">do not</b> enforce the specific constraints required by the PowerPoint application.</p><ul data-path-to-node="14" style="padding-inline-start: 32px; font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="14,0,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Constraint:</b> Microsoft PowerPoint restricts the value of the <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">firstSlideNum</code> attribute to the bounds <b style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">[0, 9999]</b>.</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="14,1,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Official Reference:</b> This restriction is documented in the Microsoft Open Specifications: <b style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">MS-OI29500 (Part 1, Section 19.2.1.26, presentation (Presentation))</b>.</p></li></ul><p data-path-to-node="15" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">By implementing <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">setFirstSlideNumber(int num)</code> at the <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">XMLSlideShow</code> level, POI can validate the input against the <b style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">PowerPoint application constraints</b>. This ensures that the files created by POI are guaranteed to be considered valid and correctly rendered by Microsoft Office, improving data integrity and user experience.</p><h3 data-path-to-node="16" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Testing</h3><p data-path-to-node="17" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">The feature is covered by new test methods in <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">TestXSLFSlideShow.java</code> which verify:</p><ul data-path-to-node="18" style="padding-inline-start: 32px; font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="18,0,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Default value and persistence.</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="18,1,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Successful setting of boundary values (0 and 9999).</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="18,2,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Correct behavior of <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">unsetFirstSlideNumber()</code>.</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="18,3,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Validation logic by throwing <code style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">IllegalArgumentException</code> for values outside the [0, 9999] range.</p></li></ul></body></html><!--EndFragment-->
</body>
</html>

hyun1024 committed:
Add methods to manage first slide number(firstSlideNum)

Adds methods to manage the custom starting slide number in XSLF (PowerPoint) presentations. This property is stored as the 'firstSlideNum' attribute in ppt/presentation.xml.

The following methods are added to XMLSlideShow:
- getFirstSlideNumber(): Retrieves the current starting slide number (default is 1).
- setFirstSlideNumber(int num): Sets the custom starting slide number.
- unsetFirstSlideNumber(): Removes the 'firstSlideNum' attribute, reverting to the default (1).

Constraints:
The 'set' method enforces the bounds [0, 9999] as defined by Microsoft's implementation specifications (MS-OI29500, Part 1, Section 19.2.1.26) to ensure the creation of valid PowerPoint files.

Also includes TestXSLFSlideShow updates to cover the new functionality, persistence, and validation checks.

pjfanning commented on poi-ooxml/src/main/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java:
```diff
@@ -748,4 +748,48 @@ String importBlip(String blipId, POIXMLDocumentPart parent, POIXMLDocumentPart t
+     * Subsequent slides will be numbered sequentially from this number.
+     * The default value is 1 if the property is not set in the presentation file.
+     *
+     * @return The starting number for the first slide (default is 1).
```
can you add `@since 6.0.0`?

pjfanning replied:
to all new public methods

hyun1024 replied:
@pjfanning Got it. I've added `@since 6.0.0` to all new public methods in the latest commit.

pjfanning commented on poi-ooxml/src/test/java/org/apache/poi/xslf/usermodel/TestXSLFSlideShow.java:
```diff
@@ -16,10 +16,6 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 ==================================================================== */
 package org.apache.poi.xslf.usermodel;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
```
keep these instead of using wildcards

hyun1024 replied:
@pjfanning Fixed the imports in the test file. Switched back to explicit static imports instead of wildcards.

hyun1024 committed:
Review: Apply review feedback to firstSlideNumber feature

- Add @since 6.0.0 Javadoc tag to getFirstSlideNumber(), setFirstSlideNumber(), and unsetFirstSlideNumber() methods in XMLSlideShow.
- Refactor TestXSLFSlideShow to use explicit static JUnit imports (e.g., assertEquals, assertThrows) instead of wildcard imports, adhering to project coding style guidelines.

pjfanning committed:
whitespace

pjfanning commented:
merged - thanks
